src/SessionSetup/RetrieveInstancesWidgetTest.cpp

```diff
@@ -358,4 +383,113 @@ TEST_F(RetrieveInstancesWidgetTestStarted, ReloadFails) {
+  EXPECT_EQ(project_combo_box_->currentIndex(), 1);
+
+  // The combobox content looks like the following (> marks selection):
+  // "Default Project (Test Project 1)"
+  // > "Test Project 1 (default)"
+  // "Test Project 2"
+
+  EXPECT_CALL(mock_retrieve_instances_,
+              LoadInstances(std::optional<Project>(std::nullopt),
+                            InstanceListScope(InstanceListScope::kOnlyOwnInstances)))
+      .WillOnce(Return(Future<ErrorMessageOr<QVector<Instance>>>(kTestInstancesProject1)));
+
+  project_combo_box_->setCurrentIndex(0);
+  QCoreApplication::processEvents();
+
+  EXPECT_EQ(loading_started_spy_.count(), 2);
+  EXPECT_EQ(loading_successful_spy_.count(), 2);
+  EXPECT_EQ(loading_failed_spy_.count(), 0);
+  EXPECT_EQ(initial_loading_failed_spy_.count(), 0);
+  VerifyLastLoadingReturnedInstanceList(kTestInstancesProject1);
+  EXPECT_EQ(GetQSettingsProject(), std::nullopt);
+  EXPECT_EQ(project_combo_box_->currentIndex(), 0);
+
+  // The combobox content looks like the following (> marks selection):
+  // > "Default Project (Test Project 1)"
+  // "Test Project 1 (default)"
+  // "Test Project 2"
+
+  EXPECT_CALL(mock_retrieve_instances_,
+              LoadInstances(std::optional<Project>(kTestProject2),
+                            InstanceListScope(InstanceListScope::kOnlyOwnInstances)))
+      .WillOnce(Return(Future<ErrorMessageOr<QVector<Instance>>>(kTestInstancesProject2)));
+
+  project_combo_box_->setCurrentIndex(2);
+  QCoreApplication::processEvents();
+
+  EXPECT_EQ(loading_started_spy_.count(), 3);
+  EXPECT_EQ(loading_successful_spy_.count(), 3);
+  EXPECT_EQ(loading_failed_spy_.count(), 0);
+  EXPECT_EQ(initial_loading_failed_spy_.count(), 0);
+  VerifyLastLoadingReturnedInstanceList(kTestInstancesProject2);
+  EXPECT_EQ(GetQSettingsProject(), kTestProject2);
+  EXPECT_EQ(project_combo_box_->currentIndex(), 2);
+
+  // The combobox content looks like the following (> marks selection):
+  // "Default Project (Test Project 1)"
+  // "Test Project 1 (default)"
+  // > "Test Project 2"
+}
+
+TEST_F(RetrieveInstancesWidgetTestStarted, ProjectChangeFailed) {
+  // Note: The project combo box ui tests do not work with QTest::mouseClick, since it not possible
+  // to click on a combobox item with the QTest framework. setCurrentIndex is used instead.
+
+  // The combobox content looks like the following (> marks selection):
+  // > "Default Project (Test Project 1)"
+  // "Test Project 1 (default)"
+  // "Test Project 2"
+
+  EXPECT_CALL(mock_retrieve_instances_,
+              LoadInstances(std::optional<Project>(kTestProject1),
+                            InstanceListScope(InstanceListScope::kOnlyOwnInstances)))
+      .WillOnce(Return(Future<ErrorMessageOr<QVector<Instance>>>(ErrorMessage{"error"})));
+
+  project_combo_box_->setCurrentIndex(1);
+
+  // close the error message box
+  QTimer::singleShot(5, []() {
+    QApplication::activeModalWidget()->close();
+    QCoreApplication::exit();
+  });
```

Comment:
Why is there a timeout needed? I'm asking because if this requires waiting then it's a potential source of flakiness.

Reply:
I found another way of doing this, with invokeMethod and Qt::QueuedConnection

Reply:
That's what I wanted to suggest when waiting is not actually needed. BTW, `QTimer::singleShot` with a timeout of `0` also works. But I believe QMetaObject::invokeMethod is more clear here.